blog/2023-02-16-do-you-need-to-build-auth/index.md

```diff
@@ -0,0 +1,124 @@
+---
+slug: do-you-need-to-build-auth
+authors: gao
+tags: [auth, dev, authentication, authorization]
+---
+
+# Do you need to build your own auth for apps?
+
+:::note Prologue
+Although article is inspired from personal dev experience and I tried my best to keep things objective, it's still non-negligible that I'm one of the creators of Logto (an auth product).
+
+I would like to point out this first for a better reading experience.
+:::
+
+I’ve seen a lot of developers are asking questions like “should I build my own auth for my app”. While the answer cannot be simple yes or no, I’d like to write an article to breakdown the implementation and demonstrate the pros and cons to help you decide.
+
+> **TL;DR** If you're still learning, auth will be a good challenge. If you want full control, it's worth building your own auth system. Otherwise, it depends on several conditions.
+
+## Intro
+
+As a developer, I have built many applications during my career. No matter the programming language, there is a common foundation I always need to construct: user auth.
+
+It was a negligible part since everything was straightforward going back to 20 years ago:
+
+- Implement a registration and sign-in system with username and password.
+- Create a mechanism to maintain a user's session.
+- About security? MD5 is the answer.
+
+That’s it. Then I could focus on “the real business”. Didn't hear about MD5? You missed out on the "good times" of software development. Nowadays, developers face more challenges when building sign-in/up.
+
+It sounds alarmist, but let me go through with an example.
+
+## Example: An online bookstore
+
+Let’s say you are trying to build an online bookstore with an API service and a web frontend app.
+
+First, the scope of “auth” should be defined. Auth can be explained as authentication and authorization, and they have totally different definitions and use cases:
+
+:::note 🤔 About Auth
+
+- Authentication (AuthN) answers the question “Which identity do you own?”
+- Authorization (AuthZ) answers the question “What can you do?”
+  :::
+
+I wrote an article [CIAM 101: Authentication, Identity, SSO](../2022-11-28-ciam-101-authn-identity-sso/index.mdx) to discuss these concepts in detail.
+
+### Choose authentication methods
+
+Let’s start with “authentication”, which is user sign-in in our example. Besides the username-and-password method, here are some trending methods people are adopting for a better user conversion and security:
```

Comment:
```suggestion
Let’s start with “authentication”, which is user sign-in in our example. Besides the username-and-password method, here are some trending methods people are adopting for better user conversion and security reasons:
```

Reply:
> for better user conversion and security

it means for better "user conversion and security"

Reply:
I just feel a bit odd to say `for security` without `reasons`.
* People would do this for security.
* People would do this for security reasons.

Well, I might just be nit-picking, both are fine actually

Reply:
People would do this for a better security.

Reply:
ok then, I'm good with it